File: boj/binary-search/1790.java
import java.io.*;
import java.util.*;

public class Main {

    static long findLength(int number) {
        long ans = 0;

        for (int start=1, len=1; start<=number; start*=10, len++) {
            int end = start*10-1;
            if (end>number) { end = number; }
            ans += (long)(end-start+1)*len;
        }
        return ans;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        String[] input = br.readLine().split(" ");

        int N = Integer.parseInt(input[0]);
        long M = Long.parseLong(input[1]);

        if (findLength(N) < M) {
            bw.write("-1");
            bw.flush();
            return;
        }

        int start = 1;
        int end = N;

        int ans = 0;
        while (start <= end) {
            int mid = (start+end)/2;
            long midLen = findLength(mid);

            if (midLen >= M) {
                end = mid-1;
                ans = mid;
            }
            else {
                start = mid+1;
            }
        }
        String str = String.valueOf(ans);
        long l = findLength(ans);
        int index = (int)(str.length()-(l-M+1));

        bw.write(str.charAt(index));
        bw.flush();
    }
}

File: boj/binary-search/1790.swift
import Foundation

class Boj1790 {
    
    func findLength(_ number: Int) -> Int64 {
        var answer = Int64()
        
        var length = 1
        for start in sequence(first: 1, next: { $0*10 })
            .prefix(while: { $0 <= number }){
                var end = start*10-1
                if end > number { end = number }
                answer += Int64((end-start+1)*length);
                length+=1
        }
        return answer
    }
    
    func main() -> String {
        guard let input = readLine()?.split(separator: " ").compactMap({Int(String($0))}) else { return "-1" }
        
        let N = Int(input[0])
        let M = Int(input[1])
        
        if findLength(N) < M { return "-1" }
        
        var start = 1
        var end = N
        
        var result = 0
        while start <= end {
            let mid = (start+end)/2
            let midLen = findLength(mid)
            
            if midLen >= M {
                end = mid-1
                result = mid
            }
            else {
                start = mid+1
            }
        }
        
        let resultString = String(result)
        let resultLength = findLength(result)
        let index = Int64(resultString.count)-(resultLength-Int64(M)+1)
        
        return resultString[Int(index)]
    }
}

extension String {
    subscript(x: Int) -> String {
        return String(self[self.index(self.startIndex, offsetBy: x)])
    }
}

print(Boj1790().main())

